sdlib/d/gc/extent.d

```diff
@@ -478,50 +574,84 @@ unittest allocfree {
 }
 
 /**
- * Packed Free Space is stored as a 15-bit unsigned integer, in one or two bytes:
+ * Packed Free Space is stored as a 14-bit unsigned integer, in one or two bytes:
  *
  * /---- byte at ptr ----\ /-- byte at ptr + 1 --\
  * B7 B6 B5 B4 B3 B2 B1 B0 A7 A6 A5 A4 A3 A2 A1 A0
- * \_________15 bits unsigned integer_________/  \_ Set if and only if B0..B7 used.
+ * \_______14 bits unsigned integer________/  \  \_ Set if and only if B0..B7 used.
+ *                                             \_ Set when finalizer is present;
+ *                                                preserved when writing.
  */
 ushort readPackedFreeSpace(ushort* ptr) {
 	auto data = loadBigEndian(ptr);
-	auto mask = 0x7f | -(data & 1);
-	return (data >> 1) & mask;
+	auto mask = 0x3f | -(data & 1);
+	return (data >> 2) & mask;
 }
 
+enum finalizerBit = nativeToBigEndian!ushort(ushort(0x2));
+
 void writePackedFreeSpace(ushort* ptr, ushort x) {
-	assert(x < 0x8000, "x does not fit in 15 bits!");
+	assert(x < 0x4000, "x does not fit in 14 bits!");
 
-	bool isLarge = x > 0x7f;
-	ushort native = (x << 1 | isLarge) & ushort.max;
+	bool isLarge = x > 0x3f;
+	ushort native = (x << 2 | isLarge) & ushort.max;
 	auto base = nativeToBigEndian(native);
 	auto mask = -isLarge | nativeToBigEndian!ushort(ushort(0xff));
 
 	auto current = *ptr;
 	auto delta = (current ^ base) & mask;
-	auto value = current ^ delta;
+	auto value = (current ^ delta) | (current & finalizerBit);
```

Comment:
Change the mask instead of fetching the finalizer bit and you'll do this for free.